core/drlibc/drlibc_x86.asm

```diff
@@ -40,8 +40,8 @@
  * See comments in x86.asm on the format here.
  */
 
-#include "../asm_defines.asm"
-#include "x86_asm_defines.asm" /* PUSHGPR, POPGPR, etc. */
+#include "../arch/asm_defines.asm"
```

Comment:
We've put arch/ and arch/x86 on the include dirs list, so I think this line is the only one that needs a fix: the other 2 below that you've changed should work fine as-is and don't need any changes?

For this one: maybe better to just remove the "../" and let the include dirs find it.  That way we don't have to change it if this file gets moved again.  WDYT?

Reply:
That is true, but apparently there is different path in CMake when VisualStudio generator is used as this error is always thrown:

`drlibc/drlibc_x86.asm(43): fatal error C1083: Cannot open include file: '../asm_defines.asm': No such file or directory`
and
`drlibc/drlibc_xarch.asm(37): fatal error C1083: Cannot open include file: 'asm_defines.asm': No such file or directory`

In `cpp2asm_support.cmake` in function `add_asm_target` there is different path for VisualStudio, maybe this is the one causing issue in the first place, but without this modification VisualStudio2013 can't compile drlibc. I can investigate issue deeper, to see from where problem originates in the first place.

Reply:
I'm following this issue a bit deeper, and compared difference between VisualStudio and nmake generated build.
In case of a VisualStudio pre processing is done without include paths (eg. wrong one), while in case on nmake pre processing is done with correct include path.

Here is diff between 2 versions:
`C:\PROGRA~2\MICROS~1.0\VC\bin\cl.exe /nologo $(ASM_FLAGS) $(ASM_INCLUDES) $(ASM_DEFINES) -DCPP2ASM -E /EP drlibc_xarch.asm > drlibc_xarch.asm.obj.s`

while in case of VisualStudio pre processing looks like:

`cl.exe  /nologo /I C:/dynamorio/buildvc -E /EP drlibc/drlibc_x86.asm > C:/dynamorio/buildvc/core/drlibc_x86.asm.s"`

This is the only difference between these 2 builds, and yes, seems that include path is missing in this particular case.

Reply:
Here is the line in cpp2asm_support.cmake:
`
COMMAND "${CMAKE_C_COMPILER}" ARGS ${CMAKE_CPP_FLAGS}
        /I "${PROJECT_BINARY_DIR}" ${extra_defs}
        -E ${CPP_NO_LINENUM} "${source}"
        ">" "${s_file}"`

This one doesn't specify all includes for pre processing. Only project_binary_dir, that is why files can't be found.

Reply:
replacing above with this:

`COMMAND "${CMAKE_C_COMPILER}" ARGS ${CMAKE_CPP_FLAGS} 
        -DCPP2ASM
        /I${CMAKE_CURRENT_SOURCE_DIR}/arch 
        /I ${CMAKE_CURRENT_SOURCE_DIR}/arch/${ARCH_NAME}
        /I "${PROJECT_BINARY_DIR}" ${extra_defs}
        -E ${CPP_NO_LINENUM} "${source}"
        ">" "${s_file}"`

I'm not wizard of CMake, is there any way to get include directories at this point, and add them to command line + -DCPP2ASM?

Now VisualStudio is able to compile without any issues, and no modifications in asm files is required.

Reply:
https://github.com/deroko/dynamorio/commit/37d590994511ec1a1d09d46d12ae2ca0042f0de2 something like this should do the trick, without modifying asm files.

Reply:
This could solve the issue with VisualStudio: https://github.com/deroko/dynamorio/commit/2b148e12236e23c9c833b1844d62fd6d07481371 

If there is better solution to get include_directories, this would solve VisualStudio issue in a nicer way.

Reply:
In core/CMakeLists.txt at line 1131 we add all these dirs using CMake's `include_directories` which I believe adds them to a global list used in all targets.  But if there is something weird with the VS generators (probably related to CMAKE_ASM_COMPILE_OBJECT) I'm ok with just putting your patch in, if we don't want to spend more time investigating: probably not worth it.  I'll just merge your patch.

Reply:
I think that modifying cpp2asm_support.cmake for Visual Studio only is a better solution than changing asm files, as issue is now identified properly :)

Reply:
If you want to look further into that, ok, I will not do the merge.

Reply:
Thank you, I will continue looking into it to make proper fix.

Reply:
I think this is the better one : https://github.com/deroko/dynamorio/commit/5764006848b2fc2609d7b78af13c1ab22ed3786b

There seems not to be a way to get include_directories at this stage. One way would be to move in `core/CMakeListst.txt` `include_directories` from 1131 to line 70, but that would break building. Maybe there is some way in VS build to expand AdditionalIncludeDirectories in CustomBuild step, but tbh, I don't know how.

Reply:
> One way would be to move in core/CMakeListst.txt include_directories from 1131 to line 70, but that would break building. 

Why does that break the build?

Reply:
It can't find os_exports.h as OSNAME is defined later on (line 319 in `core/CMakeLists.txt`). If I start changing CMakeLists.txt in core just to fix this issue in "nicer" way (eg. to have INCLUDE_DIRECTORIES in cpp2asm_support.cmake), I'm afraid that I will break something in code which is already working perfectly fine, and would stick with adding only code in cpp2asm which is required to find asm includes during preprocessing.

Reply:
The problem with listing all the include dirs again in cpp2asm is that now we have a maintenance burden: we have to update both places.  If we go this route, please put comments in both places reminding future developers to keep them in synch.

Reply:
I know, that is also why I tried to deal with INCLUDE_DIRECTORIES :( I've added these comments and fixes in this branch : https://github.com/deroko/dynamorio/commits/vs2013build-fix2 . You can chose which one is a better solution (both work).

Reply:
You mean https://github.com/DynamoRIO/dynamorio/compare/master...deroko:vs2013build-fix or https://github.com/DynamoRIO/dynamorio/compare/master...deroko:vs2013build-fix2?  Given the use of dot dot in many other places I would go with the smaller diff that just adds dot dot.

Reply:
The one already merged it the one with dot dot fix. Only thing missing is maybe adding -DCPP2ASM in cpp2asm_support.cmake (like is done for all other targets), but as far as I can see this define passed on command line is not used at all. Merged one should do the job for VisualStudio builds.

Reply:
OK I will merge.